### PR TITLE
Load controller from action for AMD version

### PR DESCRIPTION
**Pull request by purusho**
This PR will fix the loading of controllers for the AMD version.
Sample App's PR : https://github.com/rendrjs/rendr-examples/pull/15

### CODE REVIEW

**Comment by ingiulio** [comment]
+1

File: shared/base/router.js [hunk 1]
```diff
@@ -166,7 +168,21 @@ _.extend(BaseRouter.prototype, Backbone.Events, {
             callback.apply(this, arguments);
           }
         }
-        realAction.call(this, params, next);
+        // in AMD environment realAction is the string containing path to the controller
+        // which will be loaded async (might be preloaded)
+        // Only used in AMD environment
+        if (isAMDEnvironment) {
+          requireAMD([realAction], function(controller) {
```
Comment (alexindigo) [inline shared/base/router.js]: Looks like `realAction` if a function (object) and `requireAMD` expects list of strings as dependencies.

Comment (purusho) [inline shared/base/router.js]: No, for an AMD app we are getting a string : https://github.com/purusho/rendr/blob/master/shared/base/router.js#L89

Comment (alexindigo) [inline shared/base/router.js]: Then let's check for the variable type, which will protect us from changes down the road
and will allow more flexible treatment of AMD environment on the server.

**Comment by alexindigo** [comment]
:+1: 

@rendrjs/maintainers LGTM

**Comment by saponifi3d** [comment]
Seems good - can you please add a few tests around this?

**Comment by purusho** [comment]
test added.

File: shared/base/router.js [hunk 2]
```diff
@@ -184,6 +199,11 @@ _.extend(BaseRouter.prototype, Backbone.Events, {
   },
 
   /**
+   * exposing for mocking in test
+   */
+  _requireAMD: require,
```
Comment (alexindigo) [inline shared/base/router.js]: Have you tried to run actual app with this change?

Comment (purusho) [inline shared/base/router.js]: yes, and it's working.

Comment (alexindigo) [inline shared/base/router.js]: Cool :)

File: test/shared/base/router.test.js [hunk 3]
```diff
@@ -162,4 +166,70 @@ describe('BaseRouter', function () {
...
+        var requireSpy = sinon.spy();
+
+        var requireAMD = sinon.stub(router, '_requireAMD').returns(requireSpy);
+
+        router.route('/home', 'home#index');
+
+        var getHandlerCall = getHandler.getCall(0);
+
+        // call action function
+        getHandlerCall.args[0]({}, callback);
```
Comment (alexindigo) [inline test/shared/base/router.test.js]: Btw, check this out:

```
spy.callArg(argNum)
Like yield, but with an explicit argument number specifying which callback to call. Useful if a function is called with more than one callback, and simply calling the first callback is not desired.

"calling the last callback": function () {
    var callback = sinon.stub();
    callback(function () {
        console.log("Success!");
    }, function () {
        console.log("Oh noes!");
    });

    callback.callArg(1); // Logs "Oh noes!"
}


spy.callArgWith(argNum, [arg1, arg2, ...])
```

From here: http://sinonjs.org/docs/#stubs

Comment (purusho) [inline test/shared/base/router.test.js]: Thanks, changed to use  `spy.callArgWith`

**Comment by saponifi3d** [comment]
Aw yeah! 83% coverage! We're gettin' there!

**Pull request merged**